stan/math/rev/fun/elt_multiply.hpp

```diff
@@ -33,11 +34,10 @@ auto elt_multiply(const Mat1& m1, const Mat2& m2) {
     arena_t<promote_scalar_t<var, Mat2>> arena_m2 = m2;
     arena_t<ret_type> ret(arena_m1.val().cwiseProduct(arena_m2.val()));
     reverse_pass_callback([ret, arena_m1, arena_m2]() mutable {
-      for (Eigen::Index i = 0; i < arena_m2.size(); ++i) {
-        const auto ret_adj = ret.adj().coeffRef(i);
-        arena_m1.adj().coeffRef(i) += arena_m2.val().coeff(i) * ret_adj;
-        arena_m2.adj().coeffRef(i) += arena_m1.val().coeff(i) * ret_adj;
-      }
+      auto ret_adj = ret.adj().array();
+      eigen_results(arena_m1.adj(), arena_m2.adj()) = eigen_expressions(
+          arena_m1.adj().array() + arena_m2.val().array() * ret_adj,
+          arena_m2.adj().array() + arena_m1.val().array() * ret_adj);
```

Comment:
I'm not sure how I feel about this. What if we did something like tag dispatching?

```cpp
      eigen_results(std::tuple<std::plus, std::plus>(), arena_m1.adj(), arena_m2.adj()) = eigen_expressions(
           arena_m2.val().array() * ret_adj,
           arena_m1.val().array() * ret_adj);
```

Reply:
The common case is going to be `+=` so what if we just did that by default and users would just do a minus in `eigen_expression`?

Reply:
I dont like tag dispatch. It is verbose. But I agreee += will be used a lot. I will add an option for `eigen_results() += eigen_expressions()`.

Reply:
This might be kind of silly but what if we had `operator+-=`, `operator-+=` for these? That looks kind of wacky, I can think of few situations where we do more than two. 

```cpp
      eigen_results(arena_m1.adj(), arena_m2.adj()) +-= eigen_expressions(
           arena_m2.val().array() * ret_adj,
           arena_m1.val().array() * ret_adj);
```

Though maybe that's two weird and a nicer thing is just to have a minus on the expr you want like

```cpp
      eigen_results(arena_m1.adj(), arena_m2.adj()) += eigen_expressions(
           arena_m2.val().array() * ret_adj,
           -(arena_m1.val().array() * ret_adj));
```

Reply:
AFAIK there is no such thing as operator+-= in c++. So we have to go with the second option.